packages/puppeteer-core/src/common/bidi/Frame.ts

```diff
@@ -207,4 +211,229 @@ export class BidiFrame extends Frame {
     this.sandboxes[MAIN_SANDBOX][Symbol.dispose]();
     this.sandboxes[PUPPETEER_SANDBOX][Symbol.dispose]();
   }
+
+  #bindings = new Set<string>();
+  override async exposeFunction<Args extends unknown[], Ret>(
+    name: string,
+    fn: (...args: Args) => Awaitable<Ret>
+  ): Promise<void> {
+    if (this.#bindings.has(name)) {
+      throw new Error(
+        `Failed to add page binding with name ${name}: globalThis['${name}'] already exists!`
+      );
+    }
+    this.#bindings.add(name);
+
+    const argsChannel = `__puppeteer__${this._id}_page_exposeFunction_${name}_args`;
+    const resolveChannel = `__puppeteer__${this._id}_page_exposeFunction_${name}_resolve`;
+    const rejectChannel = `__puppeteer__${this._id}_page_exposeFunction_${name}_reject`;
+
+    const connection = this.#context.connection;
+
+    type SendArgsChannel = (value: [id: number, args: Args]) => void;
+    type SendResolveChannel = (
+      value: [id: number, resolve: (ret: FlattenHandle<Awaited<Ret>>) => void]
+    ) => void;
+    type SendRejectChannel = (
+      value: [id: number, reject: (error: unknown) => void]
+    ) => void;
+
+    interface RemotePromiseCallbacks {
+      resolve: Deferred<Bidi.Script.RemoteValue>;
+      reject: Deferred<Bidi.Script.RemoteValue>;
+    }
+
+    const bindingMaps = new Map<string, Map<number, RemotePromiseCallbacks>>();
+
+    const handleArgumentsMessage = async (
```

Comment:
let's extract this code into a class and make these inner methods and types members of the class.

Reply:
+1 for this change.